### PR TITLE
Added IndexStoreExtension interface

### LINKED ISSUE #325: [FEATURE] Create IndexStoreExtension interface
**Is your feature request related to a problem?**

Extension points (see https://github.com/opensearch-project/opensearch-sdk-java/issues/315) will be (optionally) implemented on Extensions. Eventually these points may need to be communicated to OpenSearch to be registered in the appropriate service or module.

Allowing extensions to inherit from appropriate interfaces depending on their type will match the current Plugin implementation.

**What solution would you like?**

Add a new interface IndexStoreExtension with default implementations for extension points matching IndexStorePlugin.java.

### PR DESCRIPTION
**Pull request by nassipkali**
### Description
This pr adds IndexStoreExtension

### Issues Resolved
#325 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/IndexStoreExtension.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package org.opensearch.sdk;
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/IndexStoreExtension.java:1]: Needs a license header.  Running `./gradlew spotlessApply` should add one automatically

File: src/main/java/org/opensearch/sdk/IndexStoreExtension.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import org.opensearch.index.IndexSettings;
+import org.opensearch.index.shard.ShardPath;
+import org.opensearch.indices.recovery.RecoveryState;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A plugin that provides alternative directory implementations.
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/IndexStoreExtension.java:16]: Replace `A plugin` with `An extension` in all the documents

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Thanks for your contribution!  A few documentation tweaks needed.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
Also please add some unit tests in `TestExtensionInterfaces` for the methods with default return values.

It's possible to test interfaces and their abstract methods with some mock/verify stuff, but I'm not sure that's necessary.

**Comment by dbwiddis** [comment]
And, note that your commits need a DCO signoff. :)

**Comment by nassipkali** [comment]
> Also please add some unit tests in `TestExtensionInterfaces` for the methods with default return values.
> 
> It's possible to test interfaces and their abstract methods with some mock/verify stuff, but I'm not sure that's necessary.

are my tests correct?

**Review by ryanbogan — APPROVED** [review]

**Comment by ryanbogan** [comment]
Looks like it needs a rebase with main to address merge conflicts

**Comment by nassipkali** [comment]
Sorry for review request flood. I thought that a review can be requested from several people at once

**Comment by dbwiddis** [comment]
> Sorry for review request flood. I thought that a review can be requested from several people at once

Heh, no worries, we already get tons of email so this is barely a blip.  But everyone's already requested, you can re-request by clicking the little refresh icon next to someone who has reviewed.

Screenshot 2023-03-14 at 11 27 35 AM

File: src/test/java/org/opensearch/sdk/TestExtensionInterfaces.java [hunk 3]
```diff
@@ -96,7 +99,6 @@ void testESearchExtension() {
         assertTrue(searchExtension.getIndexSearcherExecutorProvider().isEmpty());
     }
 
-    @Test
```
Comment (dbwiddis) [inline src/test/java/org/opensearch/sdk/TestExtensionInterfaces.java:99]: This shouldn't be removed.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
LGTM except for the incorrectly-deleted test annotation!

**Review by dbwiddis — APPROVED** [review]
LGTM!

**Review by ryanbogan — APPROVED** [review]

**Comment by ryanbogan** [comment]
@nassipkali Thanks for the contribution!

**Pull request merged**